source/extensions/request_id_utils/uuid/uuid_impl.cc

```diff
@@ -0,0 +1,107 @@
+#include "extensions/request_id_utils/uuid/uuid_impl.h"
+
+#include <cstdint>
+#include <string>
+
+#include "envoy/http/header_map.h"
+
+#include "common/common/utility.h"
+#include "common/runtime/runtime_impl.h"
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace RequestIDUtils {
+
+void UUIDUtils::setRequestID(Http::RequestHeaderMap& request_headers) {
+  // TODO(PiotrSikora) PERF: Write UUID directly to the header map.
+  std::string uuid = random.uuid();
+  ASSERT(!uuid.empty());
+  request_headers.setRequestId(uuid);
+}
+
+void UUIDUtils::ensureRequestID(Http::RequestHeaderMap& request_headers) {
+  if (!request_headers.RequestId()) {
+    setRequestID(request_headers);
+  }
+}
+
+void UUIDUtils::maybePreserveRequestIDInResponse(Http::ResponseHeaderMap& response_headers,
+                                                 const Http::RequestHeaderMap& request_headers) {
+  if (request_headers.EnvoyForceTrace() && request_headers.RequestId()) {
```

Comment:
Alternatively check `request_headers.EnvoyForceTrace()` can be kept outside of this implementation. In that case we add a configuration option in separate pull request to allow always sending request ID header back even without forced tracing. This way we will be able omit "maybe" part of the method and keep if logic outside.

It's generally useful for correlating requests with client-side logs even if tracing itself is disabled for this particular request.

Reply:
Agreed that makes sense to me and has been a requested feature before.